### PR TITLE
Configure Molecule in Travis CI

### PR DESCRIPTION
**Pull request by vcrhonek**
Configures Molecule to do basic checks (lint/syntax, run, idempotence, Ansible check mode functionality) of the role with default parameters (tests/tests_default.yml) using Travis CI and centos6/7 docker containers.

Because Travis was already used by network role, please check mainly .travis.yml, whether it makes sense to incorporate molecule into it in this way.

### CODE REVIEW

**Comment by pcahyna** [comment]
`Could not find a version that satisfies the requirement flake8<4,>=3.6.0 (from molecule) (from versions: 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.3.1, 1.4, 1.5, 1.6, 1.6.1, 1.6.2, 1.7.0, 2.0, 2.1.0, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.3.0, 2.4.0, 2.4.1, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.5.4, 2.5.5, 2.6.0, 2.6.1, 2.6.2, 3.0.0b1, 3.0.0b2, 3.0.0, 3.0.1, 3.0.2.dev0, 3.0.2.dev1, 3.0.2, 3.0.3, 3.0.4, 3.1.0, 3.1.1, 3.2.0, 3.2.1, 3.3.0.dev0, 3.3.0, 3.4.0, 3.4.1, 3.5.0)
No matching distribution found for flake8<4,>=3.6.0 (from molecule)
The command "pip install tox tox-travis molecule docker" failed and exited with 1 during .`

**Comment by vcrhonek** [comment]
> Travis fails with: `Could not find a version that satisfies the requirement flake8<4,>=3.6.0 (from molecule) (from versions: 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.3.1, 1.4, 1.5, 1.6, 1.6.1, 1.6.2, 1.7.0, 2.0, 2.1.0, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.3.0, 2.4.0, 2.4.1, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.5.4, 2.5.5, 2.6.0, 2.6.1, 2.6.2, 3.0.0b1, 3.0.0b2, 3.0.0, 3.0.1, 3.0.2.dev0, 3.0.2.dev1, 3.0.2, 3.0.3, 3.0.4, 3.1.0, 3.1.1, 3.2.0, 3.2.1, 3.3.0.dev0, 3.3.0, 3.4.0, 3.4.1, 3.5.0) No matching distribution found for flake8<4,>=3.6.0 (from molecule) The command "pip install tox tox-travis molecule docker" failed and exited with 1 during .`

I see, that happens only with Python 2.6, other jobs are fine. But it causes skip of tox testing which is unfortunate.

**Comment by tyll** [comment]
> > Travis fails with: `Could not find a version that satisfies the requirement flake8<4,>=3.6.0 (from molecule) (from versions: 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.3.1, 1.4, 1.5, 1.6, 1.6.1, 1.6.2, 1.7.0, 2.0, 2.1.0, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.3.0, 2.4.0, 2.4.1, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.5.4, 2.5.5, 2.6.0, 2.6.1, 2.6.2, 3.0.0b1, 3.0.0b2, 3.0.0, 3.0.1, 3.0.2.dev0, 3.0.2.dev1, 3.0.2, 3.0.3, 3.0.4, 3.1.0, 3.1.1, 3.2.0, 3.2.1, 3.3.0.dev0, 3.3.0, 3.4.0, 3.4.1, 3.5.0) No matching distribution found for flake8<4,>=3.6.0 (from molecule) The command "pip install tox tox-travis molecule docker" failed and exited with 1 during .`
> 
> I see, that happens only with Python 2.6, other jobs are fine. But it causes skip of tox testing which is unfortunate.

IMHO it would be nicer to add the molecule commands to the tox config and  install the dependencies there. Then we can run the tests only for certain python versions. I will take a look.

**Comment by vcrhonek** [comment]
> > > Travis fails with: `Could not find a version that satisfies the requirement flake8<4,>=3.6.0 (from molecule) (from versions: 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.3.1, 1.4, 1.5, 1.6, 1.6.1, 1.6.2, 1.7.0, 2.0, 2.1.0, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.3.0, 2.4.0, 2.4.1, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.5.4, 2.5.5, 2.6.0, 2.6.1, 2.6.2, 3.0.0b1, 3.0.0b2, 3.0.0, 3.0.1, 3.0.2.dev0, 3.0.2.dev1, 3.0.2, 3.0.3, 3.0.4, 3.1.0, 3.1.1, 3.2.0, 3.2.1, 3.3.0.dev0, 3.3.0, 3.4.0, 3.4.1, 3.5.0) No matching distribution found for flake8<4,>=3.6.0 (from molecule) The command "pip install tox tox-travis molecule docker" failed and exited with 1 during .`
> > 
> > 
> > I see, that happens only with Python 2.6, other jobs are fine. But it causes skip of tox testing which is unfortunate.
> 
> IMHO it would be nicer to add the molecule commands to the tox config and install the dependencies there. Then we can run the tests only for certain python versions. I will take a look.

Thanks, that would be great as I have no experience with tox yet.

**Comment by tyll** [comment]
I tested this locally and noticed that `molecule test` fails with this error message:

```
    failed: [localhost] (item={'name': 'centos7', 'image': 'vcrhonek/centos7-systemd', 'volumes': ['/sys/fs/cgroup:/sys/fs/cgroup:ro'], 'privileged': True}) => {"changed": false, "checksum": "0a234f1e797e8263d3a567b965259ff9bc2bf42b", "item": {"image": "vcrhonek/centos7-systemd", "name": "centos7", "privileged": true, "volumes": ["/sys/fs/cgroup:/sys/fs/cgroup:ro"]}, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}
```

**Comment by tyll** [comment]
My change for tox support are here: https://github.com/tyll/network/tree/molecule

Also this change seems helpful to make the log output of molecule test useful:

```
diff --git a/molecule/default/molecule.yml b/molecule/default/molecule.yml
index 486238a..666d732 100644
--- a/molecule/default/molecule.yml
+++ b/molecule/default/molecule.yml
@@ -18,6 +18,7 @@ platforms:
     privileged: true
 provisioner:
   name: ansible
+  log: true
   lint:
     name: ansible-lint
   playbooks:
```

**Comment by tyll** [comment]
@vcrhonek how is the Dockerfile.j2 template used? When I change it to install libselinux-python, it still fails. When I remove it, it does not behave differently... Seems like it is not needed and only confusing.

**Comment by vcrhonek** [comment]
> @vcrhonek how is the Dockerfile.j2 template used? When I change it to install libselinux-python, it still fails.
> Removing makes a difference, though...

It is used to build docker container used as a test environment for role. We can get rid of it and install everything right into the docker images from molecule.yml (vcrhonek/centos7-systemd for example). See:
https://www.jeffgeerling.com/blog/2018/testing-your-ansible-roles-molecule
(search for Use pre-built Docker images with Molecule)

I don't hit libselinux-python issue locally though (without tox).

**Comment by tyll** [comment]
> I don't hit libselinux-python issue locally though (without tox).

After I found out about `molecule --debug test` I noticed that it fails because the selinux python module is not installed on the host instead of not being available in the container. I will give it another try and report back.

**Comment by tyll** [comment]
I got it working now - the centos6 image is not supported as-is and exposes an interesting problem.
It does not contain the initscripts package which contains the `service` command and service files. Without it, the service facts from Ansible do not return anything. This breaks the detection of which provider is currently active. It could be worked around with a patch like

```
diff --git a/defaults/main.yml b/defaults/main.yml
index 1a11a96..a0d72a8 100644
--- a/defaults/main.yml
+++ b/defaults/main.yml
@@ -6,7 +6,7 @@ network_connections: []
 network_provider_os_default: "{{ 'initscripts' if ansible_distribution in ['RedHat', 'CentOS'] and ansible_distribution_major_version is version('7', '<') else 'nm' }}"
 # If NetworkManager.service is running, assume that 'nm' is currently in-use,
 # otherwise initscripts
-network_provider_current: "{{ 'nm' if 'NetworkManager.service' in ansible_facts.services and ansible_facts.services['NetworkManager.service']['state'] == 'running' else 'initscripts' }}"
+network_provider_current: "{{ 'nm' if 'NetworkManager.service' in ansible_facts.services | default([]) and ansible_facts.services['NetworkManager.service']['state'] == 'running' else 'initscripts' }}"
 # Default to the auto-detected value
 network_provider: "{{ network_provider_current }}"
```

But this would also mask any other errors with service facts. To me this seems to be an artificial problem with this docker image that users would not hit in the wild. Therefore I think the best approach would be to use a docker image that contains initscripts already for CentOS 6.
It might even be a good idea to also have two CentOS 7 images, one with NetworkManager running and one without to cover both cases and have a test that checks the provider detection is working as intended. What are your opinions?

**Comment by vcrhonek** [comment]
> But this would also mask any other errors with service facts. To me this seems to be an artificial problem with this docker image that users would not hit in the wild. Therefore I think the best approach would be to use a docker image that contains initscripts already for CentOS 6.

I agree with you.

> It might even be a good idea to also have two CentOS 7 images, one with NetworkManager running and one without to cover both cases and have a test that checks the provider detection is working as intended. What are your opinions?

In my opinion this makes very good sense for network role. But probably not for other linux-system-roles - if the idea is still to use the same docker images and molecule configuration for all of them.

**Comment by tyll** [comment]
> > But this would also mask any other errors with service facts. To me this seems to be an artificial problem with this docker image that users would not hit in the wild. Therefore I think the best approach would be to use a docker image that contains initscripts already for CentOS 6.
> 
> I agree with you.

Will you create a custom container image for this or how do we proceed?

> > It might even be a good idea to also have two CentOS 7 images, one with NetworkManager running and one without to cover both cases and have a test that checks the provider detection is working as intended. What are your opinions?
> 
> In my opinion this makes very good sense for network role. But probably not for other linux-system-roles - if the idea is still to use the same docker images and molecule configuration for all of them.

For other roles with multiple providers it might also make sense to test with different initial providers, for example for timesync, I guess. So maybe we need some variations anyhow,

File: molecule/default/molecule.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+lint:
+  name: yamllint
+  options:
+    config-file: molecule/default/yamllint.yml
+platforms:
+  - name: centos6
+    image: centos:6
+    privileged: true
+  - name: centos7
+    image: vcrhonek/centos7-systemd
```
Comment (tyll) [inline molecule/default/molecule.yml]: Please add the image the to linuxsystemroles docker organization and use it from there.

**Comment by pcahyna** [comment]
> > > It might even be a good idea to also have two CentOS 7 images, one with NetworkManager running and one without to cover both cases and have a test that checks the provider detection is working as intended. What are your opinions?
> > 
> > 
> > In my opinion this makes very good sense for network role. But probably not for other linux-system-roles - if the idea is still to use the same docker images and molecule configuration for all of them.
> 
> For other roles with multiple providers it might also make sense to test with different initial providers, for example for timesync, I guess. So maybe we need some variations anyhow,

I think you are overthinking it. We found many limitations of Molecule testing, so we want to use it only when it is reasonably simple to do it to collect the low-hanging fruits like static linting and basic check mode/idempotency checks. Configuring more complex checks, such as images with other providers, is IMO above the "reasonably simple" threshold. Anyway, using different system versions (centos-6, centos-7, centos-8) will lead to testing of different providers without any additional effort, because different systems have different defaults, won't it?

**Comment by vcrhonek** [comment]
> 
> Will you create a custom container image for this or how do we proceed?
> 

I've created  the image for CentOS 6 with initscripts and added it along with CentOS 7 image to  linuxsystemroles docker organization. Where do you think should Dockerfiles for them live?

I'm going to take a look at your proposed changes for tox support, thanks.

**Comment by pcahyna** [comment]
This can't be merged because of a failure:
> You need to install \"jmespath\" prior to running json_query filter 
One needs to install jmespath on the control host: (https://github.com/tyll/network/blob/a1a9450b4b33cf6d54f4bb26b37176efb5d60e8c/tox.ini#L14 )

and python2.6 tests errored: https://github.com/linux-system-roles/network/pull/107#issuecomment-483993294

**Comment by tyll** [comment]
I moved my cleanup commits to #108 and split them properly.

**Comment by pcahyna** [comment]
one Travis CI test still fails:
ERROR! Unexpected Exception, this is probably a bug: Failed to detect selinux python bindings at /usr/lib64/python3.6/site-packages
This is the Python3.6 test, others are passing.

File: .travis.yml [hunk 2]
```diff
@@ -1,3 +1,4 @@
+---
```
Comment (pcahyna) [inline .travis.yml]: see https://github.com/linux-system-roles/network/pull/108/files#r278538735

File: .travis.yml [hunk 3]
```diff
@@ -7,16 +7,15 @@ python:
...
-  - pip install tox tox-travis molecule docker
+  - pip install tox tox-travis
 
 script:
   - tox
-  - molecule --version
-  - ansible --version
-  - molecule lint
-  - molecule syntax
-  - molecule test
```
Comment (pcahyna) [inline .travis.yml]: The two blocks above seem to revert parts of b83d15f09b91fe92cde761f91f154fc4f1b13eb3, can you squash those two commits to minimize the diffs and reduce the chance of conflicts, please?

File: .travis.yml [hunk 4]
```diff
@@ -7,16 +7,15 @@ python:
   - 2.7
   - 3.6
   - 3.7-dev
+
 services:
   - docker
```
Comment (pcahyna) [inline .travis.yml]: How does Travis start Docker, if Docker installation has been moved to Tox? (below)

**Comment by pcahyna** [comment]
> > > But this would also mask any other errors with service facts. To me this seems to be an artificial problem with this docker image that users would not hit in the wild. Therefore I think the best approach would be to use a docker image that contains initscripts already for CentOS 6.
> > 
> > 
> > I agree with you.
> 
> Will you create a custom container image for this or how do we proceed?

An alternative to using our own image would be to build it in Travis from a vanilla image.

https://docs.travis-ci.com/user/docker/#building-a-docker-image-from-a-dockerfile

Would that work?

File: .travis.yml (hunk 4 above)
Comment (tyll) [reply .travis.yml]: tox installs only the docker python bindings, not the service itself.

File: .travis.yml (hunk 2 above)
Comment (tyll) [reply .travis.yml]: resolved this.

File: .travis.yml (hunk 3 above)
Comment (tyll) [reply .travis.yml]: I moved the changes to .travis.yml in one squashed commit.

**Comment by tyll** [comment]
> An alternative to using our own image would be to build it in Travis from a vanilla image.
> 
> https://docs.travis-ci.com/user/docker/#building-a-docker-image-from-a-dockerfile
> 
> Would that work?

It will take a little bit more ressources because it requires to always rebuild the image. Also it might be a bit more complicated since we need to build the two images a little bit more different (one with initscripts, the other with systemd).

**Comment by tyll** [comment]
> > Will you create a custom container image for this or how do we proceed?
> 
> I've created the image for CentOS 6 with initscripts and added it along with CentOS 7 image to linuxsystemroles docker organization. Where do you think should Dockerfiles for them live?

How about in a molecule repo that contains a sample/reference molecule configuration for all/new roles.

**Comment by tyll** [comment]
Regarding the selinux failure: Ubuntu seems to be special here compared to Fedora/CentOS... I have some changes prepared but I need to verify/upstream them. Would be great to merge #108 already, so that those changes are already done. We can also skip the travis changes here for now because the other changes already work when running it locally on Fedora.

**Comment by tyll** [comment]
It seems to be ready now to be merged. We need to update the tox config in case https://github.com/ssbarnea/selinux/pull/3 gets merged and published.

**Comment by tyll** [comment]
Updated the tox config since upstream merged my PR.

**Comment by tyll** [comment]
> Updated the tox config since upstream merged my PR.

Changed back to the custom selinux module since upstream needs to publish the change on PyPi before it can be used by tox.

**Comment by tyll** [comment]
@pcahyna @vcrhonek from my POV this is ready now. @pcahyna can you please review my yaml changes here or in #108.

**Comment by tyll** [comment]
[citest bad]

File: tox.ini [hunk 5]
```diff
@@ -13,7 +13,11 @@ deps =
     molecule_{lint,syntax,test}: docker
     molecule_{lint,syntax,test}: jmespath
     molecule_{lint,syntax,test}: molecule
-    molecule_{lint,syntax,test}: selinux
+    # The selinux pypi shim does not work with Ubuntu (as used by Travis), yet.
+    # Therefore use a fork with Ubuntu support. This can be changed once the
+    # update is available on PyPi.
+    # molecule_{lint,syntax,test}: selinux
+    molecule_{lint,syntax,test}: git+https://github.com/tyll/selinux-pypi-shim@fulllocation
```
Comment (pcahyna) [inline tox.ini]: @i386x @vcrhonek I think we can revert this part given that https://github.com/pycontribs/selinux/pull/3 is merged and released.

**Pull request merged**